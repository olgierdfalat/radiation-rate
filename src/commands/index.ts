export * from './checksum';
export * from './filterFiles';
export * from './wrongFiles';
export * from './excelExport';
