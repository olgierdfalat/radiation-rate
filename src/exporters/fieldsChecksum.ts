
export class FieldsChecksum {
  private folderPath: string;
  constructor(folderPath: string) {
    this.folderPath = folderPath;
  }
}