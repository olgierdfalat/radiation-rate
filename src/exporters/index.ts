export * from './exportDataProvider';
export * from './excel';