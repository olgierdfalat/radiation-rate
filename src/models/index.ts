export * from './fieldModel';
export * from './rowModel';
export * from './exportData';
export * from './deviceRows';
export * from './deviceDataProvider';
export * from './interrogationsFactory';
export * from './keyValuePair';
export * from './valueTypePair';