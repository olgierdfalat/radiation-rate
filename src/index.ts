import { Command } from 'commander';
import * as commands from './commands';


const command = new Command();
command
  .version('0.0.1')
  .description('CLI for converting heart valves output data into Excel format')
  .option('-c, --checksum <device>', 'generates checksums of all log files for specific device type e.g.: StJude, Biotronik or Medtronik', commands.checksum)
  .option('-e, --excluded <device>', 'generates list of excluded log files for specific device type e.g.: StJude, Biotronik or Medtronik', commands.excludedFiles)
  .option('-i, --included <device>', 'generates list of included log files for specific device type e.g.: StJude, Biotronik or Medtronik', commands.includedFiles)
  .option('-w, --invalid <device>', 'generates list of wrong log files for specific device type e.g.: StJude, Biotronik or Medtronik', commands.invalidFiles)
  .parse(process.argv);

