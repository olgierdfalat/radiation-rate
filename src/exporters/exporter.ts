import { includedFiles } from './../commands/filterFiles';
import {DeviceRows} from '../models/deviceRows';
import {ExportData} from '../models/exportData';
import * as interrogations from './../interrogations';


export class Exporter {
  private folderPath: string;
  private device: string;
  constructor(folderPath: string, device: string) {
    this.folderPath = folderPath;
    this.device = device;
  }

  protected async getRowsForDevices(): Promise<DeviceRows[]> {
    const files = await includedFiles(this.device, false);
    const devicesRows: DeviceRows[] = [];
    for (let i = 0; i < files.length; i++) {
      const file = files[i];
      const interrogation = new interrogations.StJude(file.filePath); // TODO: resolve dynamically
      const data = await interrogation.getData();
      const deviceId = data.getDeviceId();
      const findIndex = devicesRows.findIndex(fg => fg.deviceId == deviceId);
      if (findIndex != -1) {
        devicesRows[findIndex].rows.push(data.getRow());
      }
      else {
        devicesRows.push({deviceId, rows: [data.getRow()]});
      }
    }

    return devicesRows;
  }

  protected async getExportData(): Promise<ExportData[]> {
    const result: ExportData[] = [];
    const devicesRows = await this.getRowsForDevices();
    const dataProvider = new interrogations.ExportDataProvider();
    for (let i = 0; i < devicesRows.length; i++) {
      const deviceRows = devicesRows[i];
      const exportData = dataProvider.getExportData(deviceRows.deviceId, deviceRows.rows);
      result.push(exportData);
    }
    return result;
  }
}