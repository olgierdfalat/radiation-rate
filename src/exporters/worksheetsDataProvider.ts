import * as models from '../models';
import checksum from '../util/checksum';

export class WorksheetsDataProvider {
  getWorksheetData(deviceId: string, rows: Array<models.WorksheetRow>): models.WorksheetData   {
    const columnsLength = [...new Set([].concat.apply([], rows.map(x => x.map(f => f.name))))].length;
    const data: models.WorksheetData = {name: 'interrogations', deviceId, columns: new Array(columnsLength), rows: []};
    const rowsChecksums: string[] = new Array(rows.length);

    for (let i = 0; i < rows.length; i++) {
      const row = rows[i];
      const newRow = new Array<models.ValueTypePair>(columnsLength);

      for (let j = 0; j < row.length; j++) {
        const field = row[j];
        const columnIndex = data.columns.indexOf(field.name);
        if (columnIndex != -1) {
          newRow[columnIndex] = {value: field.value, type: field.type};
        }
        else {
          const firstEmptyColumnIndex = data.columns.findIndex(c => !c);
          data.columns[firstEmptyColumnIndex] = field.name;
          newRow[firstEmptyColumnIndex] = {value: field.value, type: field.type};
        }
      }

      const rowChecksum = checksum(newRow.map(i => (i.value || '') + ':' + (i.type || '')).join('|'));
      if (!rowsChecksums.includes(rowChecksum)) {
        data.rows.push(newRow);
      }
      rowsChecksums[i] = rowChecksum;
    }
    return data;
  }
}