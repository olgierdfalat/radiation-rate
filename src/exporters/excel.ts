import { Exporter } from './exporter';

export class Excel extends Exporter  {
   constructor(folderPath: string, device: string) {
     super(folderPath, device);
   }
}