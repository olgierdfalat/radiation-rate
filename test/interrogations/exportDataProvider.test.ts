import {ExportDataProvider} from './../../src/exporters/exportDataProvider';
import * as models from '../../src/models';

describe('Interrogations data provider class', () => {
  it('gets export data - removes duplicates and distributes values into correct columns', () => {
    const exportDataProvider = new ExportDataProvider();
    const inputData = [
                        ['column1:value1','column2:value2', 'column3:value3'],
                        ['column3:value3','column2:value2', 'column1:value1'],
                        ['column2:value2','column3:value3', 'column1:value1'],
                        ['column4:value4','column5:value5'],
                        ['column6:value6'],
                        ['column3:value3','column2:value2', 'column1:value1', 'column7:value7'],
                        ['column1:aaa','column2:bbb', 'column5:ccc', 'column7:ddd']
                      ]
    const rows: models.StJudeRowModel[] = inputData.map(item => {
        const fields: models.KeyValuePair[] = item.map(field => {
          const [key, value] = field.split(':');  
          return {key, value};
        })
        const row:models.StJudeRowModel = fields.map((field, index) => {
          return {name: field.key, id: index, type: 'string', firstValue: field.value, secondValue: null};
        })
        
        return row;
    });
    
    expect(exportDataProvider.getWorksheetData('1234', rows)).toMatchSnapshot();
  });
})