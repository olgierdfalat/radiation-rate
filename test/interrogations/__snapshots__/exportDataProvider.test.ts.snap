// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Interrogations data provider class gets export data - removes duplicates and distributes values into correct columns 1`] = `
Object {
  "columns": Array [
    "column1",
    "column2",
    "column3",
    "column4",
    "column5",
    "column6",
    "column7",
  ],
  "deviceId": "1234",
  "rows": Array [
    Array [
      Object {
        "type": "string",
        "value": "value1",
      },
      Object {
        "type": "string",
        "value": "value2",
      },
      Object {
        "type": "string",
        "value": "value3",
      },
      undefined,
      undefined,
      undefined,
      undefined,
    ],
    Array [
      undefined,
      undefined,
      undefined,
      Object {
        "type": "string",
        "value": "value4",
      },
      Object {
        "type": "string",
        "value": "value5",
      },
      undefined,
      undefined,
    ],
    Array [
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
      Object {
        "type": "string",
        "value": "value6",
      },
      undefined,
    ],
    Array [
      Object {
        "type": "string",
        "value": "value1",
      },
      Object {
        "type": "string",
        "value": "value2",
      },
      Object {
        "type": "string",
        "value": "value3",
      },
      undefined,
      undefined,
      undefined,
      Object {
        "type": "string",
        "value": "value7",
      },
    ],
    Array [
      Object {
        "type": "string",
        "value": "aaa",
      },
      Object {
        "type": "string",
        "value": "bbb",
      },
      undefined,
      undefined,
      Object {
        "type": "string",
        "value": "ccc",
      },
      undefined,
      Object {
        "type": "string",
        "value": "ddd",
      },
    ],
  ],
}
`;
